src/Framework/Analyzer/src/DelegateEndpoints/DetectMismatchedParameterOptionality.cs

```diff
@@ -0,0 +1,137 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers.DelegateEndpoints;
+
+public partial class DelegateEndpointAnalyzer : DiagnosticAnalyzer
+{
+    internal const string DetectMismatchedParameterOptionalityRuleId = "ASP0006";
+
+    private static void DetectMismatchedParameterOptionality(
+        in OperationAnalysisContext context,
+        IInvocationOperation invocation,
+        IMethodSymbol methodSymbol)
+    {
+        var value = invocation.Arguments[1].Value;
+        if (value.ConstantValue is not { HasValue: true } constant ||
+            constant.Value is not string routeTemplate)
+        {
+            return;
+        }
+
+        var parametersInArguments = methodSymbol.Parameters;
+        var parametersInRoute = GetParametersFromRoute(routeTemplate);
+
+        foreach (var parameter in parametersInArguments)
+        {
+            var isOptional = parameter.IsOptional || parameter.NullableAnnotation != NullableAnnotation.NotAnnotated;
+            var location = parameter.DeclaringSyntaxReferences.SingleOrDefault()?.GetSyntax().GetLocation();
```

Comment:
Ah, good case to think through!

 So I added a test scenario to handle this and it looks like the declaring reference will always point to the location of the declaration method and not the definition method. 

This is (probably) fine for the analyzer, but it presents a problem for the codefix where we want to fix up both the declaration and definition, otherwise the user will get a `CS8611` warning at build-time. It won't present any issues at runtime since the nullability context also only looks at the parameter nullability in the declaration.

Soooooo, I think this is fine to have in the analyzer as long as I can figure out a way to get the method definition updated with the correct nullability annotations in the codefix.

Reply:
So, I ended up changing the approach for this. It _is_ possible to all partial representations of a method associated with the parameter in the codefix but it involves pulling out the symbols from the semantic models. So, it's a lot more forward to just do the lookup in the analyzer and then use the locations computed from that.

To support this, I adapted some code from Roslyn for retrieving the declaration/definition part of each partial method and iterating through that.

This means that for partial methods, users will receive two warnings on both the declaration and definition and the codefix will address both.